### ISSUE TITLE
msgpack-json conversion produces invalid json strings

### ISSUE THREAD

**Issue by schrewe** [taocpp/json#96]
I fuzzed the implemented conversions with [CI-Fuzz](https://www.code-intelligence.com/). Using a fuzz test that checks the msgpack-json conversion roundtrip property I discovered that there are inputs for which the conversion from msgpack string to json string does not produce a valid json string. You can find the fuzz test code [here](https://gist.github.com/schrewe/8235d10359c3532e3cff1a1c6ed23a10) 

The input

0000  95 F2 24 7C 65 EC                                ò$|eì
    
triggers the assert in line 50 of the fuzz test.
That means that the conversion of input from msgpack to a json string (line 24-27) works, but when trying to convert this json string back to msgpack in the lines 32-34 a parse error exception is thrown. That means that s1 is not a valid json string.
This is a bug since the result of lines 24-27 is to my understanding intended to be a json string. I would expect a exception be thrown there instead.

If I can help investigating this bug in any way feel free to ask.

**Comment by ColinH** [comment]
The problem has nothing to do with parsing as shown by the content of the exception which is `msgpack requires array size`.

What's going on is the following. The taoJSON Events interface has *two* pairs of functions `begin_array`/`end_array` (and the same for objects), one that does not take an argument and one that takes the number of sub-values (array elements and object member, respectively).

The idea is that in the case where the caller already knows the size of an array (or object) in advance it should call the functions *with* the size as argument since this might make things easier and more efficient for the callee.

What you are triggering here is an incompatibility between generating Events from a JSON string representation, where the size is *not* known in advance, wherefore the functions without size are called, and consuming Events by the MsgPack encoder which *requires* the sizes since that format, unlike, say, CBOR, does not have an "open ended" encoding of arrays where a terminator is used to signal the end; MsgPack requires the size to be encoded at the beginning of the array (or object).

For efficiency reasons, i.e in order to not hide some significant runtime complexity, this "parity mismatch" is not fixed automatically. However there is an easy manual work-around that I also just tested by adapting your source:

You change the data type of `s1` and `s3` from a `std::string` to a `tao::json::value` by using `tao::json::events::to_value` as Events consumer instead of `tao::json::events::to_string`. With the thereby obtained `tao::json::value` in-memory representation you can use `tao::json::events::from_value` as Events producer which *is* compatible with `tao::json::msgpack::events::to_string` as Events consumer since the former uses `std::vector` for arrays and therefore always has the size available...

I'm not quite sure whether this messes up other aspects of your testing approach, let me know if this is the case and we can talk about it in detail. Otherwise I hope this helps.

**Comment by schrewe** [comment]
Thank you for the detailed explanation, I was not aware of this implementation detail and missed the hint in the exception message.
I don't think this messes up the testing approach, it was only a mistake I made when writing the fuzz target because of a lack of knowledge about the right functions to call.

**Comment by schrewe** [comment]
@ColinH, I changed the fuzz target to implement the workaround you suggested. I hope I got evrything right this time.
https://gist.github.com/schrewe/3bfa16a7ccc7fe9fb5d7d04fbb5eb5da
As I understood you, this should be the correct way of calling the functions/API.
I have a additional question: Would inputs to this fuzz target need to be sanitized / checked before in some way? Or is the code appropriate to be used to parse untrusted data?
